src/UCTNode.cpp

```diff
@@ -220,32 +221,43 @@ UCTNode* UCTNode::uct_select_child(int color) {
     auto parentvisits = size_t{0};
     for (const auto& child : m_children) {
         if (child->valid()) {
-            parentvisits += child->get_visits();
-            if (child->get_visits() > 0) {
+            if (!child->first_visit()) {
+                parentvisits += child->get_visits();
                 total_visited_policy += child->get_score();
+            } else if (child->get_score() > best_unvisited_psa) {
+                best_unvisited_psa = child->get_score();
+                best = child.get();
             }
         }
     }
 
+    if (parentvisits == 0) {
+        assert(best != nullptr);
+        return best;
+    }
+
     auto numerator = std::sqrt((double)parentvisits);
-    auto fpu_reduction = cfg_fpu_reduction * std::sqrt(total_visited_policy);
-    // Estimated eval for unknown nodes = original parent NN eval - reduction
-    auto fpu_eval = get_net_eval(color) - fpu_reduction;
+    auto best_value = -DBL_MAX;
+
+    if (best_unvisited_psa > -DBL_MAX) {
+        // Estimated eval for unknown nodes = original parent NN eval - reduction
+        auto fpu_reduction = cfg_fpu_reduction * std::sqrt(total_visited_policy);
+        auto fpu_eval = get_net_eval(color) - fpu_reduction;
+        auto best_unvisited_puct = cfg_puct * best_unvisited_psa * numerator;
+        best_value = fpu_eval + best_unvisited_puct;
+    }
 
     for (const auto& child : m_children) {
-        if (!child->active()) {
```

Comment:
Any reason for switching from active() to valid()? This would remove the benefit of root child pruning.

Reply:
Whoops! That was a merge mistake..
So that means I have to add checking whether unvisited nodes are active in the first loop too. Ok.